java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -434,22 +433,40 @@ private static boolean isLocalMethodInvocation(MethodInvocationTree tree) {
   }
 
   private void resetFieldValues(MethodInvocationTree tree) {
-    boolean changed = false;
-    Map<Symbol, SymbolicValue> values = Maps.newHashMap(programState.values);
-    for (Map.Entry<Symbol, SymbolicValue> entry : values.entrySet()) {
-      Symbol symbol = entry.getKey();
-      if (isField(symbol)) {
-        VariableTree variable = ((Symbol.VariableSymbol) symbol).declaration();
-        if (variable != null) {
-          changed = true;
-          SymbolicValue nonNullValue = constraintManager.supersedeSymbolicValue(variable);
-          values.put(symbol, nonNullValue);
+    final PMap<Symbol, SymbolicValue>[] values = new PMap[]{programState.values};
+    programState.values.forEach(new PMap.Consumer<Symbol, SymbolicValue>() {
+      @Override
+      public void accept(Symbol symbol, SymbolicValue value) {
+        if (isField(symbol)) {
+          VariableTree variable = ((Symbol.VariableSymbol) symbol).declaration();
+          if (variable != null) {
+            SymbolicValue nonNullValue = constraintManager.supersedeSymbolicValue(variable);
+            values[0] = values[0].put(symbol, nonNullValue);
+          }
         }
       }
+    });
+    if(programState.values != values [0]) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)